module/zfs/vdev_disk.c

```diff
@@ -415,6 +415,8 @@ vdev_disk_dio_put(dio_request_t *dr)
 		vdev_disk_dio_free(dr);
 
 		if (zio) {
+			vdev_disk_t *vd = zio->io_vd->vdev_tsd;
+			atomic_dec(&vd->vd_dr);
```

Comment:
May want to use `atomic_add_unless()` here to protect against `vd_dr` becoming negative.

Reply:
if(atomic_dec_if_positive(&vd->vd_dr) <= 0)
maybe set return code here, so it returns proper error?